virtcontainers/qemu.go

```diff
@@ -833,67 +833,93 @@ func (q *qemu) hotplugVFIODevice(device *config.VFIODev, op operation) error {
 	return nil
 }
 
-func (q *qemu) hotplugMacvtap(drive *VethEndpoint) error {
+func (q *qemu) hotAddNetDevice(name, id, hardAddr string, VMFds, VhostFds []*os.File) (string, error) {
 	var (
 		VMFdNames    []string
 		VhostFdNames []string
 	)
-	for i, VMFd := range drive.NetPair.VMFds {
+	for i, VMFd := range VMFds {
 		fdName := fmt.Sprintf("fd%d", i)
-		err := q.qmpMonitorCh.qmp.ExecuteGetFD(q.qmpMonitorCh.ctx, fdName, VMFd)
-		if err != nil {
-			return err
+		if err := q.qmpMonitorCh.qmp.ExecuteGetFD(q.qmpMonitorCh.ctx, fdName, VMFd); err != nil {
+			return "", err
 		}
 		VMFdNames = append(VMFdNames, fdName)
 	}
-	for i, VhostFd := range drive.NetPair.VhostFds {
+	for i, VhostFd := range VhostFds {
 		fdName := fmt.Sprintf("vhostfd%d", i)
-		err := q.qmpMonitorCh.qmp.ExecuteGetFD(q.qmpMonitorCh.ctx, fdName, VhostFd)
-		if err != nil {
-			return err
+		if err := q.qmpMonitorCh.qmp.ExecuteGetFD(q.qmpMonitorCh.ctx, fdName, VhostFd); err != nil {
+			return "", err
 		}
 		VhostFdNames = append(VhostFdNames, fdName)
 	}
-	return q.qmpMonitorCh.qmp.ExecuteNetdevAddByFds(q.qmpMonitorCh.ctx, "tap", drive.NetPair.Name, VMFdNames, VhostFdNames)
+	if err := q.qmpMonitorCh.qmp.ExecuteNetdevAddByFds(q.qmpMonitorCh.ctx, "tap", name, VMFdNames, VhostFdNames); err != nil {
+		return "", err
+	}
+
+	addr, bridge, err := q.addDeviceToBridge(id)
+	if err != nil {
+		return "", err
+	}
+	pciAddr := fmt.Sprintf("%02x/%s", bridge.Addr, addr)
+
+	devID := "virtio-" + id
+	if err = q.qmpMonitorCh.qmp.ExecuteNetPCIDeviceAdd(q.qmpMonitorCh.ctx, name, devID, hardAddr, addr, bridge.ID, int(q.config.NumVCPUs)); err != nil {
+		return pciAddr, err
+	}
+
+	return pciAddr, nil
 }
 
-func (q *qemu) hotplugNetDevice(drive *VethEndpoint, op operation) error {
+func (q *qemu) hotRemoveNetDevice(name, id string) error {
```

Comment:
This is a global comment about the factorization you're doing in `virtcontainers/qemu.go`. Please create a first commit introducing the factorization independently from the addition of `TapEndpoint`, and then you can add `TapEndpoint` with a following commit. This way, it's clearly defined what's done because of the factorization vs what's done by the addition of `TapEndpoint`.

Reply:
Done